### PR TITLE
Enable ASO

### LINKED ISSUE #3523: Expose ASO metrics in Tilt's Prometheus
This is already mostly done at https://github.com/nojnhuh/cluster-api-provider-azure/tree/aso and strictly depends on #3520.

### LINKED ISSUE #3527: Enable ASO
This issue tracks "turning on" ASO from within CAPZ. Before then, ASO changes will be integrated in a way such that ASO is not usable so progress can be made iteratively while working toward feature parity with CAPZ's current feature set. The following list tracks reaching that parity. Once the prerequisites are met, one atomic change will be made to "turn on" ASO.

This depends on only at least one of the Azure service implementations being converted to ASO. Other services will still be using the Azure SDK when this issue is completed.

```[tasklist]
### Dependencies
- [ ] #3520
- [ ] #3521
- [ ] #3522
- [ ] #3524
- [ ] #3525
- [ ] #3526
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3795
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3732
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3743
```

### LINKED ISSUE #3532: Collect ASO resources/logs in e2e artifacts
```[tasklist]
### Dependencies
-  [ ] #3527
```

### PR DESCRIPTION
**Pull request by nojnhuh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This PR enables all of the currently-merged ASO functionality which was previously inaccessible to users. Specifically, this change will install ASO alongside CAPZ and use it to manage resource groups.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3527, fixes #3523, fixes #3532

**Special notes for your reviewer**:

This PR is a work-in-progress opened early to start getting test signal and accumulate changes from other contributors as necessary. Reviews are welcome but I plan to rebase/force-push this branch frequently before removing the [WIP] status.

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [X] adds unit tests

**Release note**:

```release-note
Azure Service Operator is now installed alongside CAPZ and manages resource groups
```

### CODE REVIEW

File: azure/services/aso/tags.go
Comment (nojnhuh) [inline azure/services/aso/tags.go:1]: The commit from #3691 is included here until that PR merges. The other commits comprise the actual enablement changes.

**Comment by nojnhuh** [comment]
/test pull-cluster-api-provider-azure-e2e
^ this passed but took ~30m longer than usual. Seems like an etcd fluke, but running again to confirm.

/test pull-cluster-api-provider-azure-apiversion-upgrade
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-optional

**Comment by nojnhuh** [comment]
`pull-cluster-api-provider-azure-e2e-optional` failing because ASO auth is missing UserAssignedMSI (#3743) and WorkloadIdentity (#3732).

/test pull-cluster-api-provider-azure-apiversion-upgrade
^ VM failed provisioning on the old version before upgrade.

/test pull-cluster-api-provider-azure-e2e
^ Passed before, but at least this one didn't take any longer.

/test pull-cluster-api-provider-azure-capi-e2e
^ Looks like another VM provisioning failure.

**Comment by Jont828** [comment]
@nojnhuh Is this ready for review or would you like some more time to hack away at this?

**Comment by nojnhuh** [comment]
> @nojnhuh Is this ready for review or would you like some more time to hack away at this?

Feel free to leave any feedback. I'm not expecting to make any significant changes to this, but there's some orthogonal work that will land in other PRs that this PR depends on. So I'll continue to rebase/force-push this branch until I remove the [WIP].

**Comment by nojnhuh** [comment]
Pardon me, just flake hunting 🔍 

/test pull-cluster-api-provider-azure-apiversion-upgrade
/test pull-cluster-api-provider-azure-e2e
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e

File: controllers/helpers.go
Comment (Jont828) [inline controllers/helpers.go:610]:
```suggestion
	return groups.New(clusterScope).ShouldDeleteIndividualResources(ctx)
```

Since `groups.New()` doesn't return an error

File: docs/book/src/topics/aso.md
Comment (Jont828) [inline docs/book/src/topics/aso.md:11]:
```suggestion
## Primary changes
```

File: docs/book/src/topics/aso.md
Comment (Jont828) [inline docs/book/src/topics/aso.md:14]:
```suggestion
that may affect specific use cases are described below.
```

**Review by Jont828 — COMMENTED** [review]
I know this is still WIP but here are some initial thoughts.

**Comment by nojnhuh** [comment]
This should be ready for review, so I'm removing the [WIP] here. Trading that for a hold to make sure #3743 is implemented before merging this, but I expect those changes not to overlap with these.

/retitle Enable ASO
/hold

File: azure/services/groups/spec.go
Comment (nojnhuh) [reply azure/services/groups/spec.go:40]: I'm slightly inclined not to change these since what these functions do is already defined here and copy-pasting any more details for each type is sure to get out of sync: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/7eacff73fbfa1ac2c9a786c7ad290caca110905b/azure/interfaces.go#L137-L148

File: azure/services/groups/groups_test.go
Comment (nojnhuh) [reply azure/services/groups/groups_test.go:47]: The kind of error doesn't really matter to these tests, just whether it's nil/non-nil.

**Comment by nojnhuh** [comment]
Just pushed a couple of commits to bring back the old `groups` service and wire up some logic to preserve the old behavior when a cluster is using `UserAssignedMSI` auth since that isn't yet supported by ASO (#3743).

/test pull-cluster-api-provider-azure-e2e-optional
^ Private cluster test should pass now

File: Tiltfile
Comment (CecileRobertMichon) [inline Tiltfile:256]: why is this change needed?

File: azure/services/asogroups/groups.go
Comment (CecileRobertMichon) [inline azure/services/asogroups/groups.go:62]: this won't be possible yet since we're having to keep both packages side by side for now for UserAssigned identity but thoughts on removing the "aso" prefix everywhere in naming and going back to just "groups" as service name once we remove legacy groups?

File: config/aso/credentials.yaml
Comment (CecileRobertMichon) [reply config/aso/credentials.yaml:8]: uh is that behavior we expect? if not is there an issue tracking this in ASO to make them optional?

File: docs/book/src/topics/aso.md
Comment (CecileRobertMichon) [inline docs/book/src/topics/aso.md:1]: thanks for adding docs! 📖

**Review by CecileRobertMichon — COMMENTED** [review]
Overall lgtm

/cc @matthchr

File: Tiltfile
Comment (nojnhuh) [reply Tiltfile:256]: This was necessary to make sure that variable is still set since it was referenced by the ASO global settings Secret for fallback credentials. Now that those are all blank though I think this change can be reverted.

File: azure/services/asogroups/groups.go
Comment (nojnhuh) [reply azure/services/asogroups/groups.go:62]: Yes, that's the plan.

File: config/aso/credentials.yaml
Comment (nojnhuh) [reply config/aso/credentials.yaml:8]: I don't remember testing this very thoroughly since https://github.com/Azure/azure-service-operator/issues/2725 was fixed, but it looks like we can probably omit this Secret entirely.

Comment (nojnhuh) [reply config/aso/credentials.yaml:8]: Actually I see this error when the Secret doesn't exist at all, so I think we need to keep this with blank values. I'm following up in that PR fixing that issue: https://github.com/Azure/azure-service-operator/pull/2758#issuecomment-1691944668. That seems like a bug but I'll keep this as-is for now until we have a fix.

Comment (nojnhuh) [reply config/aso/credentials.yaml:8]: > this error

An event actually:
```
89s         Warning   Failed                  pod/azureserviceoperator-controller-manager-84446f4877-z8jw7    Error: secret "aso-controller-settings" not found
```

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

File: docs/book/src/topics/aso.md
Comment (matthchr) [inline docs/book/src/topics/aso.md:18]: Is this a permanent behavior or at some future date BYO ASO might be supported?

**Review by matthchr — APPROVED** [review]
Had one small question but overall this looks reasonable to me.

File: docs/book/src/topics/aso.md
Comment (nojnhuh) [reply docs/book/src/topics/aso.md:18]: We haven't yet discussed any plans to support BYO ASO. I think the main roadblock to that is that `clusterctl` expects certain labels applied to the CRDs for things like `clusterctl move` to work. In theory, if users add those themselves to the ASO CRDs, then things should work, but that would require a fair bit of testing that we haven't done yet.

And I think we would need `clusterctl init` to support some conditional logic that could enable/disable the ASO install alongside CAPZ to support existing users relying on ASO being installed that way.

Comment (CecileRobertMichon) [reply docs/book/src/topics/aso.md:18]: Do we have an issue to discuss BYO ASO?

Comment (matthchr) [reply docs/book/src/topics/aso.md:18]: I'm not necessarily saying you need one until such time as a user asks about it - just something worth thinking about ahead of time as it is possible it will come up.

Comment (nojnhuh) [reply docs/book/src/topics/aso.md:18]: Opened #3900 to track

File: azure/scope/managedcontrolplane.go
Comment (willie-yao) [inline azure/scope/managedcontrolplane.go:751]: If using legacy groups, would we need both Tags Specs? I may be misunderstanding here but it looks like we will have two copies of the additional tags here.

File: controllers/azurecluster_reconciler_test.go
Comment (willie-yao) [inline controllers/azurecluster_reconciler_test.go:353]: This may be out of scope for this PR but it could be worth adding some test cases for `UseLegacyGroups: true` to cover all cases.

**Review by willie-yao — COMMENTED** [review]
Looks great to me @nojnhuh! Just a few minor questions from my end.

File: azure/scope/managedcontrolplane.go
Comment (nojnhuh) [reply azure/scope/managedcontrolplane.go:751]: One tags spec is for the resource group and one is for the managed cluster. With ASO, tags for a resource are reconciled alongside the rest of the resource instead of with the separate `tags` service which these serve as input to. So for an ASO resource, we don't need to define a tags spec.

File: controllers/azurecluster_reconciler_test.go
Comment (nojnhuh) [reply controllers/azurecluster_reconciler_test.go:353]: I did try adding some but it turned out to be super messy and I couldn't justify fighting through that since ideally we'll drop that flag entirely in the next week or so. I can come back to these if we think it will stick around much longer than that.

**Comment by nojnhuh** [comment]
I squashed most of this but left two commits so we can revert using ASO for resource groups independently from installing ASO altogether if need be.

/hold cancel

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Comment by nojnhuh** [comment]
Quota issue should be fixed.
/retest

**Pull request merged**
